lib/SILGen/SILGenDestructor.cpp

```diff
@@ -219,6 +220,135 @@ void SILGenFunction::destroyClassMember(SILLocation cleanupLoc,
+
+  // Collect all stored properties that would form a recursive structure,
+  // so we can remove the recursion and prevent the call stack from
+  // overflowing.
+  llvm::SmallSetVector<VarDecl*, 4> recursiveLinks;
+  for (VarDecl *vd : cd->getStoredProperties()) {
+    auto Ty = vd->getInterfaceType()->getOptionalObjectType();
+    if (Ty && Ty->getCanonicalType() == SelfTy->getCanonicalType()) {
+      recursiveLinks.insert(vd);
+    }
+  }
+
+  // NOTE: Right now we only optimize linear recursion, so if there is more than one link,
+  // clear out the set and don't perform any recursion optimization.
+  if (recursiveLinks.size() > 1) {
+    recursiveLinks.clear();
+  }
+
+  return recursiveLinks;
+}
+
+
+void SILGenFunction::emitRecursiveChainDestruction(ManagedValue selfValue,
+                                                   ClassDecl *cd,
+                                                   SmallSetVector<VarDecl*, 4> recursiveLinks,
+                                                   CleanupLocation cleanupLoc) {
+  auto SelfTy = F.mapTypeIntoContext(cd->getDeclaredInterfaceType());
+
+  assert(recursiveLinks.size() <= 1 && "Only linear recursion supported.");
+
+  auto SelfTyLowered = getTypeLowering(SelfTy).getLoweredType();
+  for (VarDecl* vd : recursiveLinks) {
+    SILBasicBlock* cleanBB = createBasicBlock();
+    SILBasicBlock* noneBB = createBasicBlock();
+    SILBasicBlock* notUniqueBB = createBasicBlock();
+    SILBasicBlock* uniqueBB = createBasicBlock();
+    SILBasicBlock* someBB = createBasicBlock();
+    SILBasicBlock* loopBB = createBasicBlock();
+
+    // var iter = self.link
+    // self.link = nil
+    auto Ty = getTypeLowering(F.mapTypeIntoContext(vd->getInterfaceType())).getLoweredType();
+    auto optionalNone = B.createOptionalNone(cleanupLoc, Ty);
+    SILValue varAddr =
+      B.createRefElementAddr(cleanupLoc, selfValue.getValue(), vd,
+                             Ty.getAddressType());
+    auto iterAddr = B.createAllocStack(cleanupLoc, Ty);
+    SILValue addr = B.createBeginAccess(
+      cleanupLoc, varAddr, SILAccessKind::Modify, SILAccessEnforcement::Static,
+      false /*noNestedConflict*/, false /*fromBuiltin*/);
+    SILValue iter = B.createLoad(cleanupLoc, addr, LoadOwnershipQualifier::Copy);
+    B.createStore(cleanupLoc, optionalNone, addr, StoreOwnershipQualifier::Assign);
+    B.createEndAccess(cleanupLoc, addr, false /*is aborting*/);
+    B.createStore(cleanupLoc, iter, iterAddr, StoreOwnershipQualifier::Init);
+
+    B.createBranch(cleanupLoc, loopBB);
+
+    // while iter != nil {
+    B.emitBlock(loopBB);
+    SILValue operand = B.createLoad(cleanupLoc, iterAddr, LoadOwnershipQualifier::Copy);
+    auto operandCopy = B.createCopyValue(cleanupLoc, operand);
+    auto operandAddr = B.createAllocStack(cleanupLoc, Ty);
+    B.createStore(cleanupLoc, operandCopy, operandAddr, StoreOwnershipQualifier::Init);
+    B.createDestroyValue(cleanupLoc, operand);
+    B.createSwitchEnumAddr(
```

Comment:
Can this be `switch_enum` instead of `switch_enum_addr`? You might not need `operandAddr` at all.

Reply:
I actually could remove the whole load store before and just do a `switch_enum_addr` on the `iterAddr` directly.